### PR TITLE
[WIP] Add WithNonVoidReturnType function

### PR DESCRIPTION
**Pull request by balazsbonis**
DO NOT MERGE.

I started working on the https://github.com/FakeItEasy/FakeItEasy/issues/648 issue, just to see if I can manage to get it done, and I got to weird exception here.

First of all, I don't know if I am taking the right approach, so feel free to tell me where should I look at.

Secondly, I added a unit test to see how my code works so far (Should_return_null_from_trapper_when_set_to_return_null in ExpressionArgumentConstraintFactoryTests.cs), and it blows up saying this:

`Expected object to be <null>, but found <NULL>.`

which I cannot wrap my head around...

Can I ask one of the FakeItEasy gurus to help me out with this? :smile:

### CODE REVIEW

**Comment by thomaslevesque** [comment]
Thanks @balazsbonis !

I'll have a look at it ASAP

File: src/FakeItEasy/Configuration/AnyCallCallRule.cs [hunk 1]
```diff
@@ -37,7 +39,9 @@ protected override bool OnIsApplicableTo(IFakeObjectCall fakeObjectCall)
         {
             return
                 this.argumentsPredicate(fakeObjectCall.Arguments) &&
-                    (this.ApplicableToMembersWithReturnType == null || this.ApplicableToMembersWithReturnType.Equals(fakeObjectCall.Method.ReturnType));
+                    (this.ApplicableToMembersWithReturnType == null 
+                    || this.ApplicableToMembersWithReturnType.Equals(fakeObjectCall.Method.ReturnType)
+                    || this.ApplicableToAllNonVoidReturnTypes);
```
Comment (thomaslevesque) [inline src/FakeItEasy/Configuration/AnyCallCallRule.cs]: Should check that the return type is not `void` when `ApplicableToAllNonVoidReturnTypes` is true

File: src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs [hunk 2]
```diff
@@ -81,7 +81,7 @@ private IArgumentConstraint GetArgumentConstraintFromExpression(Expression expre
             var trappedConstraints = this.argumentConstraintTrapper.TrapConstraints(() =>
             {
                 expressionValue = InvokeExpression(expression);
-            });
+            }) ?? new IArgumentConstraint[0];
```
Comment (thomaslevesque) [inline src/FakeItEasy/Expressions/ExpressionArgumentConstraintFactory.cs]: I don't think this should be necessary... I'll try to understand what's going on.

**Comment by thomaslevesque** [comment]
> Expected object to be <null>, but found <NULL>.

Ah, I see... `result` should be null, but it's actually an instance of `EqualityArgumentConstraint` whose `ConstraintDescription` is the string "<NULL>". It's a little confusing...

File: tests/FakeItEasy.Tests/Expressions/ExpressionArgumentConstraintFactoryTests.cs [hunk 3]
```diff
@@ -44,6 +44,20 @@ public void Should_return_constraint_from_trapper_when_available()
             result.Should().BeSameAs(constraint);
         }
 
+
+        [Test]
+        public void Should_return_null_from_trapper_when_set_to_return_null()
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Tests/Expressions/ExpressionArgumentConstraintFactoryTests.cs]: I don't think you need to add this test. The test above isn't a test for `WithReturnType`, but a test for `ExpressionArgumentConstraintFactory`.

**Comment by thomaslevesque** [comment]
> First of all, I don't know if I am taking the right approach, so feel free to tell me where should I look at.

Your approach seems good to me, and it mostly works already :+1:

**Comment by blairconrad** [comment]
Thanks, @balazsbonis. Things appear to be shaping up.
I can see you and @thomaslevesque have things well in hand.

One thing that I'd suggest for the testing side would be to add [an acceptance test as described in Contributing.md](https://github.com/FakeItEasy/FakeItEasy/blob/master/CONTRIBUTING.md#tests). An advantage of doing this early would be that the question around `Should_return_null_from_trapper_when_set_to_return_null` can probably go away. ConfigurationSpecs.cs is probably the best site for such (a) test(s) at this point…

**Comment by blairconrad** [comment]
Hey, @balazsbonis. How are things going?
There's not been much activity here, and I was just wondering if you were still interested in pursuing this issue. There's no rush or anything, just making sure.

Is there something that we can do to clear things up further or otherwise assist you?

**Comment by adamralph** [comment]
@balazsbonis we took your commit and merged it, with a slight amendment (although you are still shown as author of the commit), as part of https://github.com/FakeItEasy/FakeItEasy/pull/762 - thanks very much for your contribution! Look out for your name in the forthcoming 2.1.0 release notes! :trophy:

**Pull request closed without merging**